packages/core/src/actions/emit.ts

```diff
@@ -0,0 +1,106 @@
+import isDevelopment from '#is-development';
+import {
+  ActionArgs,
+  AnyActorScope,
+  AnyMachineSnapshot,
+  EventObject,
+  MachineContext,
+  SendExpr,
+  ParameterizedObject
+} from '../types.ts';
+
+function resolveEmit(
+  _: AnyActorScope,
+  snapshot: AnyMachineSnapshot,
+  args: ActionArgs<any, any, any>,
+  actionParams: ParameterizedObject['params'] | undefined,
+  {
+    event: eventOrExpr
+  }: {
+    event:
+      | EventObject
+      | SendExpr<
+          MachineContext,
+          EventObject,
+          ParameterizedObject['params'] | undefined,
+          EventObject,
+          EventObject
+        >;
+  }
+) {
+  if (typeof eventOrExpr === 'string') {
+    throw new Error(
+      `Only event objects may be used with emit; use emit({ type: "${eventOrExpr}" }) instead`
+    );
+  }
+  const resolvedEvent =
+    typeof eventOrExpr === 'function'
+      ? eventOrExpr(args, actionParams)
+      : eventOrExpr;
+  return [snapshot, { event: resolvedEvent }];
+}
+
+function executeEmit(
+  actorScope: AnyActorScope,
+  params: {
+    event: EventObject;
+    id: string | undefined;
+    delay: number | undefined;
+  }
+) {
+  const { event, delay, id } = params;
+  // if (typeof delay === 'number') {
+  actorScope.defer(() => {
+    actorScope.emit(event);
+  });
+  return;
+  // }
+}
+
+export interface EmitAction<
+  TContext extends MachineContext,
+  TExpressionEvent extends EventObject,
+  TParams extends ParameterizedObject['params'] | undefined,
+  TEvent extends EventObject,
+  TEmitted extends EventObject
+> {
+  (args: ActionArgs<TContext, TExpressionEvent, TEvent>, params: TParams): void;
+  _out_TEvent?: TEvent;
+  _out_TEmitted?: TEmitted;
+}
+
+/**
+ * Emits an event.
+ *
+ * @param eventType The event to emit.
+ */
+export function emit<
+  TContext extends MachineContext,
+  TExpressionEvent extends EventObject,
+  TEvent extends EventObject = TExpressionEvent,
+  TParams extends ParameterizedObject['params'] | undefined =
+    | ParameterizedObject['params']
+    | undefined,
+  TEmitted extends EventObject = EventObject
+>(
+  eventOrExpr:
+    | TEmitted
+    | SendExpr<TContext, TExpressionEvent, TParams, TEmitted, TEvent>
+): EmitAction<TContext, TExpressionEvent, TParams, TEvent, TEmitted> {
+  function emit(
+    args: ActionArgs<TContext, TExpressionEvent, TEvent>,
+    params: TParams
+  ) {
+    if (isDevelopment) {
+      throw new Error(`This isn't supposed to be called`);
+    }
+  }
+
+  emit.type = 'xstate.emit';
+  emit.event = eventOrExpr;
```

Comment:
While it is an event - we already have events that mean a different thing. What do you think about calling it differently? It gets confusing when u start saying "send an event" and "emit an event". What do you think about calling this a notification?

Reply:
Yeah, but conceptually they're still the same thing. An emitted event can be thought of as sending an event to implicit "listener actors", whose logic is simply to call a function when the event is received. Pseudocode:

```ts
function on(actor, eventType, fn) {
  const listenerActor = createListenerActor(fn);

  // the emit(...) action will send the event to all listener actors
  actor.send({
    type: 'xstate.listen',
    eventType,
    ref: listenerActor
  });

  return { ... } // subscription
}
```

In fact, we should probably model it this way in the future so we get visualization/tracing benefits etc. (maybe for v6)